src/Xml/Dom/Locator/Attribute/xmlns_attributes_list.php

```diff
@@ -4,23 +4,23 @@
 
 namespace VeeWee\Xml\Dom\Locator\Attribute;
 
-use DOMNameSpaceNode;
-use DOMNode;
+use \DOM\NameSpaceNode;
+use \DOM\Node;
 use VeeWee\Xml\Dom\Collection\NodeList;
 use VeeWee\Xml\Exception\RuntimeException;
 use function VeeWee\Xml\Dom\Locator\Xmlns\linked_namespaces;
 use function VeeWee\Xml\Dom\Predicate\is_element;
 
 /**
- * @return NodeList<DOMNameSpaceNode>
+ * @return NodeList<\DOM\NameSpaceNode>
```

Comment:
TODO : NameSpaceNode does not exist anymore: All NameSpaceNodes are regular Attrs now.

Need to find a way to list all XMLNSes, since `'./namespace::*'` xpath query does not work anymore:
https://gist.github.com/veewee/9160fe7bcf0d325fc999ca1d0127d5ab

To determine : How will the API behave with the attributes changes:

* Will attributes() return both regular and xmlns attributes
* If so, do we need a new function to only return namespace-less attributes
* ...

Reply:
Indeed `namespace::*` doesn't work right now. There's an implementation issue right now for the namespace axis in XPath queries. This is because of some peculiar rules laid out by https://www.w3.org/TR/1999/REC-xpath-19991116/#namespace-nodes

In particular, the namespace axis needs to return all in-scope namespaces for an element. However that spec link states:
> Elements never share namespace nodes: if one element node is not the same node as another element node, then none of the namespace nodes of the one element node will be the same node as the namespace nodes of another element node.

So we can't return the `DOM\Attr` corresponding to the namespace declaration (if there even is one) because we'd have to return the same attribute node for different elements. A solution is to port over DOMNameSpaceNode specifically only for XPath, as a way to get read-only information about the in-scope namespaces. I think that's a sensible way to go about this, but the (new) class will extend `DOM\Node` and should have a clear name to avoid confusion (e.g. `DOM\XPathNameSpaceInfoNode` or something like that).

Browsers have not solved this issue, and simply return the empty list when you query the namespace axis.

The DOM spec never really differentiated between namespace nodes and attributes. Namespace declarations are just a special form of attributes, so it was a bit of a mistake to let DOMNameSpaceNode be returned from `getAttributeNode` etc, but it does make some sense for the namespace axis in XPath.

Let me know what you think :slightly_smiling_face: 

To answer the API question:

> Will attributes() return both regular and xmlns attributes?

`$element->attributes` and `$element->getAttributeNames()` will include both the "normal" and "xmlns" attributes.

Reply:
I've scanned the code and currently it is only used for optimizing namespaces:
https://github.com/veewee/xml/blob/3.x/src/Xml/Dom/Configurator/optimize_namespaces.php

If browsers don't have an answer for this, then maybe we shouldn't have an answer either.

Having the namespaces as attributes now solves a lot of problems already. However, for the case above (which I'm actively using) - another solution might be to have a method that lists a map of all known namespace declarations in the document or on an element?


(The question about the attributes was more directed to myself : shall I only return the non-namespace attributes as in legace spec or all attributes as in new spec - I probably gonna go with "all" which are easy to filter by the provided API in this package)

Reply:
> another solution might be to have a method that lists a map of all known namespace declarations in the document or on an element?

Reminds me of this feature request: https://bugs.php.net/bug.php?id=63410

I can add this, e.g. `public function getInScopeNamespaces(): array;` on `DOM\Element` that returns an array where the keys are the namespace prefixes and values are the URIs as strings. Note that I can't put the DOM\Attr's as values because you can have an element in a namespace without having a corresponding xmlns attribute (the serializer then adds it for you, but the attribute doesn't actually exist). Shadowed namespace prefixes would not be included.

Since I'm going to do a follow-up RFC for new features anyway, I can include it in there.

Reply:
Sounds sufficient to me!
That would make the need for the `namespace::*` xpath pretty much zero.